File: hermod.py
#!/usr/bin/python2
# IRC Bot for Hermod Telegram and Signal Gateway. Works with sigpoller for receiving
# messages from a signal group and telegramhook for a telegram group.
# configuration in /etc/hermod.json
#
# 2019, Ruben de Groot
#
# This is the python version of what started as a perl project.

import socket
import ssl
import time
import requests
import urllib
import json

## Settings
with open('/etc/hermod.json') as f:
    cfg = json.load(f)

irc_C = socket.socket(socket.AF_INET, socket.SOCK_STREAM) #defines the socket
if cfg['irc']['UseSSL'] != 0:
    irc = ssl.wrap_socket(irc_C)
else:
    irc = irc_C

print "Establishing connection to [%s]" % (cfg['irc']['node'])
# Connect to IRC channel
irc.connect((cfg['irc']['node'], cfg['irc']['port']))
irc.setblocking(False)
irc.send("PASS %s\n" % (cfg['irc']['password']))
irc.send("USER "+ cfg['irc']['nick'] +" "+ cfg['irc']['nick'] +" "+ cfg['irc']['nick'] +" :herMod-Test\n")
irc.send("NICK "+ cfg['irc']['nick'] +"\n")
irc.send("PRIVMSG nickserv :identify %s %s\r\n" % (cfg['irc']['nick'], cfg['irc']['password']))
irc.send("JOIN "+ cfg['irc']['channel'] +"\n")


firstcall = True
f = open(cfg['signal']['file'], 'r')
g = open(cfg['telegram']['file'], 'r')

while True:
    time.sleep(1)
    URL = "https://api.telegram.org/bot" + cfg['telegram']['token'] + "/sendMessage?chat_id=" + cfg['telegram']['chat_id'] + "&text="

    # Tail log Files, forward lines to IRC
    try:
        if firstcall:
            f.seek(0,2)
            g.seek(0,2)
            firstcall = False

        sigline = f.readline()
        telline = g.readline()
        if sigline != '':
            irc.send("PRIVMSG %s :%s" % (cfg['irc']['channel'], sigline))
            print "Send to IRC: " + sigline
        if telline != '':
            irc.send("PRIVMSG %s :%s" % (cfg['irc']['channel'], telline))
            print "Send to IRC: " + telline
    except Exception as e:
        print e

    # Read new IRC messages, forward lines to Signal and Telegram
    try:
        buf = irc.recv(2040)
        
        texts = buf.splitlines()
        for text in texts:
            print text
            URL = "https://api.telegram.org/bot" + cfg['telegram']['token'] + "/sendMessage?chat_id=" + cfg['telegram']['chat_id'] + "&text="

            if text.find('PRIVMSG ' + cfg['irc']['channel']) != -1:
                nick = text.partition('!')[0].strip(':')
                msg = text.rpartition('PRIVMSG ' + cfg['irc']['channel'] + ' :')[2]
                if msg[1:7] == 'ACTION':
                    msg = "[IRC] ***" + nick + " " + msg.rpartition('ACTION')[2]
                else:
                    msg = "[IRC " + nick + "]: " + msg

                # Send to Telegram
                requests.get(URL + urllib.quote(msg))

                # Send to Signal
                h = open(cfg['signal']['infile'], "a")
                h.write(msg)
                h.write("\n")
                h.close()

                print "Send to Signal and Telegram: " + msg

            # Prevent Timeout
            if text.find('PING') != -1:
                irc.send('PONG ' + text.split() [1] + '\r\n')
    except Exception:
        continue

File: hermod.pl
#!/usr/bin/perl
#
# IRC Bot for Hermod Telegram Gateway. Works with telegramhook for receiving
# messages from a telegram group.
#
# 2019-08-17, Ruben de Groot
#
# Abandoned because of SSL problems. Use the python version instead

use strict;
use warnings;
use POE qw(Component::IRC Component::IRC::Plugin::FollowTail Component::SSLify);
use HTTP::Tiny;
use URI::Escape;
use JSON qw( decode_json );

my $cfg = "/etc/hermod.json";
open my $fh, '<', $cfg or die "error opening $cfg $!";
my $config = decode_json do { local $/; <$fh> };

my $ircname  = 'Hermod gateway from irc to telegram';
my $URL = "https://api.telegram.org/bot$config->{token}/sendMessage";

# We create a new PoCo-IRC object
my $irc = POE::Component::IRC->spawn(
   nick    => $config->{nick},
   ircname => $ircname,
   server  => $config->{ircnode},
   Port    => $config->{port},
   UseSSL  => $config->{usessl},
) or die "Oh noooo! $!";

POE::Session->create(
    package_states => [
        main => [ qw(_default _start irc_001 irc_public irc_tail_input ) ],
    ],
    heap => { irc => $irc },
);

$poe_kernel->run();

sub _start {
    my $heap = $_[HEAP];

    # retrieve our component's object from the heap where we stashed it
    my $irc = $heap->{irc};
    $irc->plugin_add( 'FollowTail' => POE::Component::IRC::Plugin::FollowTail->new(
        filename => $config->{telfile},
    ));
    $irc->yield( register => 'all' );
    $irc->yield( connect => { } );
    return;
}

sub irc_001 {
    my $sender = $_[SENDER];

    # Since this is an irc_* event, we can get the component's object by
    # accessing the heap of the sender. Then we register and connect to the
    # specified server.
    my $irc = $sender->get_heap();

    print "Connected to ", $irc->server_name(), "\n";

    # we join our channel
    $irc->yield( join => $config->{channel} );
    return;
}

sub irc_public {
  my ($sender, $who, $where, $what) = @_[SENDER, ARG0 .. ARG2];
  my $nick = ( split /!/, $who )[0];
  my @data;
  my $coin = '';
  return if ( $what =~ /forbidden words/ );
  # we relay all messages straight to telegram
  my $text = uri_escape("Msg in IRC $config->{channel} by $nick: $what");
  my $data = "chat_id=$config{chat_id}&text=$text";
  HTTP::Tiny->new->get( "$URL?$data" );
  return;
}

sub irc_tail_input {
    my ($kernel, $sender, $config->{telfile}, $input) = @_[KERNEL, SENDER, ARG0, ARG1];
    $irc->yield( privmsg => $config->{channel} => $input );
    return;
}

# We registered for all events, this will produce some debug info.
sub _default {
    my ($event, $args) = @_[ARG0 .. $#_];
    my @output = ( "$event: " );

    for my $arg (@$args) {
        if ( ref $arg eq 'ARRAY' ) {
            push( @output, '[' . join(', ', @$arg ) . ']' );
        }
        else {
            push ( @output, "'$arg'" );
        }
    }
    print join ' ', @output, "\n";
    die "I got disconnected :( :(\nKilling myself..\n" if grep(/irc_disconnect/,@output);
    return;
}

sub telepoller {
    my $reply = "Polling telegram\n";
    $irc->yield( privmsg => $config->{channel} => $reply );
    sleep 10;
}

